scipy/interpolate/regrid_python.py

```diff
@@ -0,0 +1,1104 @@
+    if p != -1: # https://github.com/scipy/scipy/blob/v1.16.2/scipy/interpolate/fitpack/fpgrre.f#L97
+        # Dx.shape[0] is the number of penalty rows; add that many zero rows
+        # so Ax_aug and Q have compatible leading dimensions for in-place QR.
+        # https://github.com/scipy/scipy/blob/v1.16.2/scipy/interpolate/fitpack/fpgrre.f#L110-L118
+        Q = np.vstack([Q, np.zeros((Dx.shape[0], Q.shape[1]), dtype=float)])
+
+    # https://github.com/scipy/scipy/blob/v1.16.2/scipy/interpolate/fitpack/fpgrre.f#L106-L175
+    # Perform in-place banded QR reduction of the x-augmented system:
+    # This orthogonalizes/eliminates along x for all RHS columns in Q simultaneously.
+    # After this, fpback can do x-direction back-substitution to
+    # get c^T (partial coeffs).
+    _dierckx.qr_reduce(Ax_aug, offset_aug_x, nc_augx, Q)
+
+    # https://github.com/scipy/scipy/blob/v1.16.2/scipy/interpolate/fitpack/fpgrre.f#L246-L253
+    # Back-substitute along x to solve the reduced system:
+    #   cT has shape (ny_data_like, ncoef_x) in this calling pattern, i.e. per y-column.
+    # The API uses:
+    #   Ax_aug, nc_augx: reduced upper structure
+    #   x_x, tx, kx, w_x: x-sample grid, knot vector, degree, and (unit) weights
+    #   Q: RHS (current)
+    T, _, _ = _dierckx.fpback(
+        Ax_aug, nc_augx, x_x,
+        Q, tx, kx, w_x,
+        Q, False
+    )
+
+    # We now want to treat the *y*-direction solve with these as the new RHS.
+    # Transpose so each column corresponds to a y-solve RHS consistently.
+    Q = np.ascontiguousarray(T.T)
+
+    # If we stacked penalty rows on the y side, pad RHS with zeros to match Ay_aug.
+    if p != -1: # https://github.com/scipy/scipy/blob/v1.16.2/scipy/interpolate/fitpack/fpgrre.f#L97
+        # https://github.com/scipy/scipy/blob/v1.16.2/scipy/interpolate/fitpack/fpgrre.f#L110-L118
+        Q = np.vstack([Q, np.zeros((Dy.shape[0], Q.shape[1]), dtype=float)])
+
+    # https://github.com/scipy/scipy/blob/v1.16.2/scipy/interpolate/fitpack/fpgrre.f#L176-L245
+    # Perform in-place banded QR reduction along y for all columns of Q.
+    _dierckx.qr_reduce(Ay_aug, offset_aug_y, nc_augy, Q)
+
+    # https://github.com/scipy/scipy/blob/v1.16.2/scipy/interpolate/fitpack/fpgrre.f#L254-L269
+    # Final back-substitution along y:
+    # Returns:
+    #   C  : coefficient matrix (orientation matches Ay/BSpline expectations here)
+    #   fp : FITPACK's internal residual metric from the y-solve
+    #        (we recompute below anyway)
+    C, _, fp = _dierckx.fpback(
+        Ay_aug, nc_augy,
+        x_y, Q, ty, ky, w_y,    # y-grid, y-knots, degree, weights
+        Q,                      # RHS -> solution becomes coefficients along y
+        False
+    )
+
+    # Build explicit (dense) design matrices to evaluate the fitted surface:
+    # _Ax: [mx × nx_coef], _Ay: [my × ny_coef]
+    _Ax = BSpline.design_matrix(x_x, tx, kx, extrapolate=False)
```

Comment:
I'm confused: `BSpline.design_matrix` constructs a CSR sparse array but the comment says dense?